File: packages/sdk/hush-go/siem/transforms/ocsf.go
package transforms

import (
	"fmt"

	"github.com/backbay-labs/clawdstrike-go/siem"
)

// OCSF Detection Finding class constants.
const (
	ocsfClassUID      = 2004 // Detection Finding (NOT deprecated 2001 Security Finding)
	ocsfCategoryUID   = 2    // Findings
	ocsfActivityCreate = 1   // Create
	ocsfTypeUID       = ocsfClassUID*100 + ocsfActivityCreate // 200401
	ocsfVersion       = "1.4.0"
)

// outcomeToOCSF maps Clawdstrike outcomes to OCSF status IDs.
var outcomeToOCSF = map[string]int{
	"allow": 1, // Success
	"deny":  2, // Failure
	"warn":  0, // Unknown
}

// severityToOCSF maps severity strings to OCSF severity IDs.
// Critical = 5 (NOT 6 which is Fatal).
var severityToOCSF = map[string]int{
	"info":     1,
	"low":      2,
	"warning":  3,
	"warn":     3,
	"medium":   3,
	"error":    4,
	"high":     4,
	"critical": 5,
}

// severityLabels maps severity IDs to OCSF label strings.
var severityLabels = map[int]string{
	0:  "Unknown",
	1:  "Informational",
	2:  "Low",
	3:  "Medium",
	4:  "High",
	5:  "Critical",
	6:  "Fatal",
	99: "Other",
}

// ToOCSF converts a SecurityEvent to OCSF v1.4.0 Detection Finding format.
func ToOCSF(event siem.SecurityEvent) map[string]interface{} {
	statusID := 0
	if id, ok := outcomeToOCSF[event.Outcome]; ok {
		statusID = id
	}

	actionID := 1     // Allowed
	dispositionID := 1 // Allowed
	switch event.Outcome {
	case "deny":
		actionID = 2      // Denied
		dispositionID = 2 // Blocked
	case "warn":
		actionID = 1       // Allowed (warn is non-blocking)
		dispositionID = 17 // Logged
	}

	ocsf := map[string]interface{}{
		"class_uid":      ocsfClassUID,
		"category_uid":   ocsfCategoryUID,
		"type_uid":       ocsfTypeUID,
		"activity_id":    ocsfActivityCreate,
		"activity_name":  "Create",
		"time":           event.Timestamp.UnixMilli(),
		"status_id":      statusID,
		"action_id":      actionID,
		"disposition_id": dispositionID,
		"message":        event.EventType,
		"metadata": map[string]interface{}{
			"version": ocsfVersion,
			"product": map[string]string{
				"name":        "ClawdStrike",
				"uid":         "clawdstrike",
				"vendor_name": "Backbay Labs",
				// Note: AgentInfo has no Version field; omit rather than use wrong value.
			},
			"original_uid": event.EventID,
		},
		"actor": map[string]interface{}{
			"user": map[string]interface{}{
				"uid":  event.Agent.ID,
				"name": event.Agent.Name,
			},
			"app_name": "clawdstrike",
		},
	}

	sevID := 0
	guardName := "unknown"
	guardMessage := event.EventType
	if event.Decision != nil {
		if id, ok := severityToOCSF[event.Decision.Severity]; ok {
			sevID = id
		}
		guardName = event.Decision.Guard
		guardMessage = event.Decision.Message
	}
	ocsf["severity_id"] = sevID
	ocsf["severity"] = severityLabels[sevID]
	ocsf["finding_info"] = map[string]interface{}{
		"uid":   event.EventID,
		"title": fmt.Sprintf("%s decision", guardName),
		"analytic": map[string]interface{}{
			"name":    guardName,
			"type_id": 1,
			"type":    "Rule",
		},
		"desc": guardMessage,
	}

	if event.Resource != nil {
		resources := []map[string]interface{}{
			{
				"type": event.Resource.Type,
				"name": resourceName(event.Resource),
			},
		}
		ocsf["resources"] = resources
	}

	return ocsf
}

func resourceName(r *siem.ResourceInfo) string {
	switch r.Type {
	case "file":
		return r.Path
	case "network":
		return r.Host
	case "tool":
		return r.Tool
	default:
		return ""
	}
}

File: packages/sdk/hush-ts/src/siem/transforms/ocsf.ts
import type { SecurityEvent } from "../types";

const OCSF_VERSION = "1.4.0";

function ocsfSeverityId(sev: SecurityEvent["decision"]["severity"]): number {
  switch (sev) {
    case "info":
      return 1; // Informational
    case "low":
      return 2;
    case "medium":
      return 3;
    case "high":
      return 4;
    case "critical":
      return 5; // Critical = 5, NOT 6 (Fatal)
    default: {
      const exhaustive: never = sev;
      return exhaustive;
    }
  }
}

function ocsfSeverityLabel(sev: SecurityEvent["decision"]["severity"]): string {
  switch (sev) {
    case "info":
      return "Informational";
    case "low":
      return "Low";
    case "medium":
      return "Medium";
    case "high":
      return "High";
    case "critical":
      return "Critical";
    default: {
      const exhaustive: never = sev;
      return exhaustive;
    }
  }
}

/**
 * Convert a SecurityEvent to an OCSF Detection Finding (class_uid=2004).
 *
 * OCSF v1.4.0 compliant with proper class_uid, category_uid, type_uid,
 * action_id, disposition_id, finding_info, and corrected severity mapping.
 */
export function toOcsf(event: SecurityEvent): Record<string, unknown> {
  const severityId = ocsfSeverityId(event.decision.severity);
  const severityLabel = ocsfSeverityLabel(event.decision.severity);
  const isWarn = event.event_type === "guard_warn";
  const actionId = isWarn ? 1 : event.decision.allowed ? 1 : 2; // 1=Allowed, 2=Denied
  const dispositionId = isWarn ? 17 : event.decision.allowed ? 1 : 2; // 17=Logged, 1=Allowed, 2=Blocked
  const statusId =
    event.outcome === "success" ? 1 : event.outcome === "failure" ? 2 : 0;

  const activityId = 1; // Create
  const classUid = 2004; // Detection Finding

  return {
    class_uid: classUid,
    category_uid: 2, // Findings
    type_uid: classUid * 100 + activityId,
    activity_id: activityId,
    activity_name: "Create",
    time: new Date(event.timestamp).getTime(),
    severity_id: severityId,
    severity: severityLabel,
    status_id: statusId,
    action_id: actionId,
    disposition_id: dispositionId,
    message: event.decision.reason,
    metadata: {
      version: OCSF_VERSION,
      product: {
        name: "ClawdStrike",
        uid: "clawdstrike",
        vendor_name: "Backbay Labs",
        version: event.agent.version,
      },
      original_uid: event.event_id,
    },
    finding_info: {
      uid: event.event_id,
      title: `${event.decision.guard} decision`,
      analytic: {
        name: event.decision.guard,
        type_id: 1, // Rule
        type: "Rule",
      },
      desc: event.decision.reason,
    },
    actor: {
      user: {
        name: event.agent.name,
        uid: event.agent.id,
      },
      app_name: "clawdstrike",
      session: event.session.id ? { uid: event.session.id } : undefined,
    },
    resources: [
      {
        name: event.resource.name,
        type: event.resource.type,
      },
    ],
  };
}
